src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -0,0 +1,1378 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.CodeAnalysis.Semantics
+{
+    internal static partial class CSharpOperationFactory
+    {
+        private static readonly ConditionalWeakTable<BoundNode, IOperation> s_cache = new ConditionalWeakTable<BoundNode, IOperation>();
```

Comment:
@CyrusNajmabadi my next work is adding parent pointer and doing better identity mechanism.

current thinking is making factory an instance owned by semantic model that have explicit cache tied to semantic model.

only thing I am not sure is whether the cache will be tied to bound node or not.